universe/Condition.cpp

```diff
@@ -9955,6 +9955,203 @@ unsigned int Not::GetCheckSum() const {
+}
+
+bool TopmostMatches::RootCandidateInvariant() const {
+    if (m_root_candidate_invariant != UNKNOWN_INVARIANCE)
+        return m_root_candidate_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->RootCandidateInvariant()) {
+            m_root_candidate_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_root_candidate_invariant = INVARIANT;
+    return true;
+}
+
+bool TopmostMatches::TargetInvariant() const {
+    if (m_target_invariant != UNKNOWN_INVARIANCE)
+        return m_target_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->TargetInvariant()) {
+            m_target_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_target_invariant = INVARIANT;
+    return true;
+}
+
+bool TopmostMatches::SourceInvariant() const {
+    if (m_source_invariant != UNKNOWN_INVARIANCE)
+        return m_source_invariant == INVARIANT;
+
+    for (auto& operand : m_operands) {
+        if (!operand->SourceInvariant()) {
+            m_source_invariant = VARIANT;
+            return false;
+        }
+    }
+    m_source_invariant = INVARIANT;
+    return true;
+}
+
+std::string TopmostMatches::Description(bool negated/* = false*/) const {
+    std::string values_str;
+    if (m_operands.size() == 1) {
+        values_str += UserString("DESC_TOPMOST_BEFORE_SINGLE_OPERAND");
+        // Pushing the negation to the enclosed conditions
+        values_str += m_operands[0]->Description(negated);
+        values_str += UserString("DESC_TOPMOST_AFTER_SINGLE_OPERAND");
+    } else {
+        values_str += UserString("DESC_TOPMOST_BEFORE_OPERANDS");
+        for (unsigned int i = 0; i < m_operands.size(); ++i) {
+            // Pushing the negation to the enclosed conditions
+            values_str += m_operands[i]->Description(negated);
+            if (i != m_operands.size() - 1) {
+                values_str += UserString("DESC_TOPMOST_BETWEEN_OPERANDS");
+            }
+        }
+        values_str += UserString("DESC_TOPMOST_AFTER_OPERANDS");
+    }
+    return values_str;
+}
+
+std::string TopmostMatches::Dump(unsigned short ntabs) const {
+    std::string retval = DumpIndent(ntabs) + "TopmostMatches [\n";
+    for (auto& operand : m_operands)
+        retval += operand->Dump(ntabs+1);
+    retval += DumpIndent(ntabs) + "]\n";
+    return retval;
+}
+
+void TopmostMatches::GetDefaultInitialCandidateObjects(const ScriptingContext& parent_context, ObjectSet& condition_non_targets) const {
+    // XXX unsure what this does
+    if (!m_operands.empty()) {
+        m_operands[0]->GetDefaultInitialCandidateObjects(parent_context, condition_non_targets); // gets condition_non_targets from first operand condition
+    } else {
+        ConditionBase::GetDefaultInitialCandidateObjects(parent_context, condition_non_targets);
+    }
+}
+
+void TopmostMatches::SetTopLevelContent(const std::string& content_name) {
+    // XXX unsure what this does
```

Comment:
need to understand if SetTopLevelContent is right, then remove comment

Reply:
`SetTopLevelContent` is used to set `m_top_level_content`, which is declared here:

https://github.com/freeorion/freeorion/blob/master/universe/ValueRef.h#L207

and which is used here:

https://github.com/freeorion/freeorion/blob/master/universe/ValueRef.cpp#L552

to return the top level bit of content in which this ValueRef is scripted. For example, in a species script, it would be the name of the species, returned as a `ValueRef::Constant<std::string>` when `ThisSpecies` is written into the script. This is helpful for writing generic macros that can be used in different contexts.

All the Effects, Conditions, and ValueRefs all have `SetTopLevelContent` functions, as they can contain eachother recursively and the final leaf ValueRef needs to have its top level content set by its enclosing object.

Reply:
Thanks for the explanation